File: core/src/main/java/org/parboiled/parserunners/TracingParseRunner.java
/*
 * Copyright (C) 2009-2011 Mathias Doenitz
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.parboiled.parserunners;

import org.parboiled.Context;
import org.parboiled.MatchHandler;
import org.parboiled.MatcherContext;
import org.parboiled.Rule;
import org.parboiled.buffers.InputBuffer;
import org.parboiled.common.*;
import org.parboiled.matchers.Matcher;
import org.parboiled.support.MatcherPath;
import org.parboiled.support.ParsingResult;
import org.parboiled.support.Position;

import static org.parboiled.common.Preconditions.checkArgNotNull;

/**
 * A {@link org.parboiled.parserunners.ParseRunner} implementation used for debugging purposes.
 * It exhibits the same behavior as the {@link ReportingParseRunner} but collects debugging information as to which
 * rules did match and which didn't.
 */
public class TracingParseRunner<V> extends ReportingParseRunner<V> implements MatchHandler {
    private Predicate<Tuple2<Context<?>, Boolean>> filter;
    private Sink<String> log;
    private MatcherPath lastPath;
    private int line;

    /**
     * Creates a new TracingParseRunner instance without filter and a console log for the given rule.
     *
     * @param rule the parser rule
     */
    public TracingParseRunner(Rule rule) {
        super(rule);
    }

    /**
     * Attaches the given filter to this TracingParseRunner instance.
     * The given filter is used to select the matchers to print tracing statements for.
     * NOTE: The given filter must be of type Predicate<Tuple2<Context<?>, Boolean>>. The reason this type is not
     * directly specified in the constructors signature is that this would make predicate expressions using the
     * {@link Predicates} operations and the predefined predicate constructors in {@link org.parboiled.support.Filters}
     * much more cumbersome to write (due to Java limited type parameters inference logic you would have to explicitly
     * state the type parameters in many places).
     *
     * @param filter the matcher filter selecting the matchers to print tracing statements for. Must be of type
     *               Predicate<Tuple2<Context<?>, Boolean>>.
     * @return this instance
     */
    @SuppressWarnings( {"unchecked"})
    public TracingParseRunner<V> withFilter(Predicate<?> filter) {
        this.filter = (Predicate<Tuple2<Context<?>, Boolean>>) checkArgNotNull(filter, "filter");
        return this;
    }

    public Predicate<Tuple2<Context<?>, Boolean>> getFilter() {
        if (filter == null) {
            withFilter(Predicates.alwaysTrue());
        }
        return filter;
    }

    /**
     * Attaches the given log to this TracingParseRunner instance.
     *
     * @param log the log to use
     * @return this instance
     */
    public TracingParseRunner<V> withLog(Sink<String> log) {
        this.log = log;
        return this;
    }

    public Sink<String> getLog() {
        if (log == null) {
            withLog(new ConsoleSink());
        }
        return log;
    }

    @Override
    protected ParsingResult<V> runBasicMatch(InputBuffer inputBuffer) {
        getLog().receive("Starting new parsing run\n");
        lastPath = null;

        MatcherContext<V> rootContext = createRootContext(inputBuffer, this, true);
        boolean matched = rootContext.runMatcher();
        return createParsingResult(matched, rootContext);
    }

    @SuppressWarnings( {"unchecked"})
    public boolean match(MatcherContext<?> context) {
        Matcher matcher = context.getMatcher();
        boolean matched = matcher.match(context);
        if (getFilter().apply(new Tuple2<Context<?>, Boolean>(context, matched))) {
            line++;
            print(context, matched); // set line-dependent breakpoint here
        }
        return matched;
    }

    private void print(MatcherContext<?> context, boolean matched) {
        Position pos = context.getInputBuffer().getPosition(context.getCurrentIndex());
        MatcherPath path = context.getPath();
        MatcherPath prefix = lastPath != null ? path.commonPrefix(lastPath) : null;
        if (prefix != null && prefix.length() > 1) getLog().receive("..(" + (prefix.length() - 1) + ")../");
        getLog().receive(path.toString(prefix != null ? prefix.parent : null));
        String line = context.getInputBuffer().extractLine(pos.line);
        getLog().receive(", " + (matched ? "matched" : "failed") + ", cursor at " + pos.line + ':' + pos.column +
                " after \"" + line.substring(0, Math.min(line.length(), pos.column - 1)) + "\"\n");
        lastPath = path;
    }
}

File: parboiled-scala/src/main/scala/org/parboiled/scala/parserunners/TracingParseRunner.scala
/*
 * Copyright (C) 2009-2011 Mathias Doenitz
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.parboiled.scala
package parserunners

import org.parboiled.parserunners.{TracingParseRunner => PTracingParseRunner}
import org.parboiled.Context
import utils.Predicate
import org.parboiled.support.MatcherPath
import org.parboiled.common.{Sink, Predicate => PPredicate, Tuple2 => T2}

/**
 * A wrapper for org.parboiled.parserunners.TracingParseRunner which returns a scala ParsingResult.
 * It provides for the ability to attach filter expressions for limiting the tracing printout to certain input and/or
 * grammar areas.
 * Note that the ParseRunner only accepts rules with zero or one value type parameter, as parsers leaving more
 * than one value on the value stack are considered to be bad style.
 */
object TracingParseRunner {
  def apply(rule: Rule0) = new TracingParseRunner[Nothing](new PTracingParseRunner[Nothing](rule))

  def apply[V](rule: Rule1[V]) = new TracingParseRunner[V](new PTracingParseRunner[V](rule))
}

class TracingParseRunner[V](val inner: PTracingParseRunner[V]) extends ParseRunner[V] {
  def run(input: Input): ParsingResult[V] = ParsingResult(inner.run(input.inputBuffer))

  /**
   * Attaches the given filter predicate limiting the tracing printout.
   * You can supply an expression consisting of either the predefined Predicates (see the Lines, Rules, Matched and
   * Mismatched objects) or custom functions with one of the following signatures as TracingPredicates:
   * Context[Any] => Boolean
   * (Context[Any], Boolean) => Boolean
   *
   * For example: "Lines(10 until 20) && Rules.below(parser.Factor)".
   */
  def filter(filter: TracingPredicate) = {
    inner.withFilter(new PPredicate[T2[Context[Any], Boolean]] {
      def apply(t: T2[Context[Any], Boolean]) = filter(t.a, t.b)
    })
    this
  }

  /**
   * Attaches the given log sink to this TracingParseRunner instance.
   */
  def log(log: Sink[String]) = {
    inner.withLog(log)
    this
  }
}

abstract class TracingPredicate extends Predicate[(Context[Any], Boolean)]

object TracingPredicate {
  implicit def fromRawPredicate(p: Predicate[(Context[Any], Boolean)]): TracingPredicate = new TracingPredicate {
    def apply(t: (Context[Any], Boolean)) = p(t)
  }

  implicit def fromContextPredicate(f: Context[Any] => Boolean): TracingPredicate = new TracingPredicate {
    def apply(t: (Context[Any], Boolean)) = f(t._1)
  }
}

object Lines {
  /**
   *  Creates a TracingPredicate which selects all lines with a number greater or equal to the given one. 
   */
  def from(firstLine: Int): TracingPredicate = toPredicate(_ >= firstLine)

  /**
   * Creates a TracingPredicate which selects all lines with a number smaller or equal to the given one.
   */
  def until(lastLine: Int): TracingPredicate = toPredicate(_ <= lastLine)

  /**
   * Creates a TracingPredicate which selects all lines in the given range of line numbers.
   */
  def apply(lineRange: Range): TracingPredicate = {
    require(lineRange.step == 1, "Ranges with step != 1 are not supported here")
    toPredicate(line => { lineRange.start <= line && line < lineRange.end })
  }

  private def toPredicate(f: Int => Boolean): TracingPredicate =
    (c: Context[Any]) => f(c.getInputBuffer.getPosition(c.getCurrentIndex).line)
}

object Rules {
  /**
   * Creates a TracingPredicate which selects only the given rules.
   */
  def only(rules: Rule*): TracingPredicate = {
    val matchers = rules.map(_.matcher)
    (c:Context[Any]) => matchers.contains(c.getMatcher)
  }

  /**
   * Creates a TracingPredicate which selects only rules on levels below the given rules.
   */
  def below(rules: Rule*): TracingPredicate = {
    val matchers = rules.map(_.matcher)
    toPredicate(path => { matchers.exists(m => (path.element.matcher ne m) && path.contains(m)) })
  }

  /**
   * Creates a TracingPredicate which selects all the given rules and their sub rules.
   */
  def apply(rules: Rule*): TracingPredicate = {
    val matchers = rules.map(_.matcher)
    toPredicate(path => { matchers.exists(path.contains) })
  }

  private def toPredicate(f: MatcherPath => Boolean): TracingPredicate = (c: Context[Any]) => f(c.getPath)
}

/**
 * A TracingPredicate selecting only rules that have matched.
 */
object Matched extends TracingPredicate {
  def apply(t: (Context[Any], Boolean)) = t._2
}

/**
 * A TracingPredicate selecting only rules that have not matched.
 */
object Mismatched extends TracingPredicate {
  def apply(t: (Context[Any], Boolean)) = !t._2
}